In [100]:
import pandas as pd

In [101]:
train = pd.read_csv("./data/train.csv")
test = pd.read_csv("./data/test.csv")
submit = pd.read_csv("./data/gender_submission.csv")

Pclass : 상위, 중간, 하위
SibSp : 형제자
Parch : 가족 관계
Embarked : 어디서 승선했는지

In [102]:
train.shape

(891, 12)

In [103]:
train.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [104]:
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [105]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# age nan값 채우기

In [106]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [107]:
# age nan값 채우기
train['Age'].fillna(train['Age'].median(), inplace=True)
test['Age'].fillna(train['Age'].median(), inplace=True)

/tmp/ipykernel_10578/688278764.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train['Age'].fillna(train['Age'].median(), inplace=True)
/tmp/ipykernel_10578/688278764.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True

In [108]:
train.shape

(891, 12)

In [109]:
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [110]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [111]:
test['Fare'].fillna(train['Fare'].median(), inplace=True)

/tmp/ipykernel_10578/3907805922.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  test['Fare'].fillna(train['Fare'].median(), inplace=True)


# Cabin nan값 채우기

In [112]:
train["Cabin"].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [113]:
test["Cabin"].unique()

array([nan, 'B45', 'E31', 'B57 B59 B63 B66', 'B36', 'A21', 'C78', 'D34',
       'D19', 'A9', 'D15', 'C31', 'C23 C25 C27', 'F G63', 'B61', 'C53',
       'D43', 'C130', 'C132', 'C101', 'C55 C57', 'B71', 'C46', 'C116',
       'F', 'A29', 'G6', 'C6', 'C28', 'C51', 'E46', 'C54', 'C97', 'D22',
       'B10', 'F4', 'E45', 'E52', 'D30', 'B58 B60', 'E34', 'C62 C64',
       'A11', 'B11', 'C80', 'F33', 'C85', 'D37', 'C86', 'D21', 'C89',
       'F E46', 'A34', 'D', 'B26', 'C22 C26', 'B69', 'C32', 'B78',
       'F E57', 'F2', 'A18', 'C106', 'B51 B53 B55', 'D10 D12', 'E60',
       'E50', 'E39 E41', 'B52 B54 B56', 'C39', 'B24', 'D28', 'B41', 'C7',
       'D40', 'D38', 'C105'], dtype=object)

In [114]:
train["Cabin_First"] = train["Cabin"].apply(lambda x : str(x)[0])

In [115]:
test["Cabin_First"] = test["Cabin"].apply(lambda x : str(x)[0])

In [116]:
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Cabin_First  
0        0         A/5 21171   7.2500   NaN        S           n  
1        0          PC 17599  71.2833   C85        C           C  
2        0  STON/O2. 3101282   7.9250   NaN        S           n  
3        0            113803  53.1000  C123        S           C  
4        0            373450   8.0500   NaN        S           n  
..     ...               ...      ...   ...      ...         ...  
886      0            211536  13.0000   NaN        S           n  
887      0            112053  30.0000   B42        S           B  
888      2        W./C. 6607  23.4500   NaN        S           n  
889      0            111369  30.0000  C148        C           C  
890      0            370376   7.7500   NaN        Q           n  

[891 rows x 13 columns]

In [117]:
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  \
0      male  34.5      0      0              330911    7.8292   NaN        Q   
1    female  47.0      1      0              363272    7.0000   NaN        S   
2      male  62.0      0      0              240276    9.6875   NaN        Q   
3      male  27.0      0      0              315154    8.6625   NaN        S   
4    female  22.0      1      1             3101298   12.2875   NaN        S   
..      ...   ...    ...    ...                 ...       ...   ...      ...   
413    male  28.0      0      0           A.5. 3236    8.0500   NaN        S   
414  female  39.0      0      0            PC 17758  108.9000  C105        C   
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S   
416    male  28.0      0      0              359309    8.0500   NaN        S   
417    male  28.0      1      1                2668   22.3583   NaN        C   

    Cabin_First  
0             n  
1             n  
2             n  
3             n  
4             n  
..          ...  
413           n  
414           C  
415           n  
416           n  
417           n  

[418 rows x 12 columns]

In [118]:
# 성에 따라 다를까 했지만 성의 값이 value_counts가 1등이 9로 미미하므로 변경 x
train["First_Name"] = train["Name"].apply(lambda x : x.split(",")[0])

In [119]:
train = train.drop(columns=["Name", "Ticket"])

In [120]:
test = test.drop(columns=["Name", "Ticket"])

In [121]:
train = train.drop(columns=["First_Name"])

In [122]:
train

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin  \
0              1         0       3    male  22.0      1      0   7.2500   NaN   
1              2         1       1  female  38.0      1      0  71.2833   C85   
2              3         1       3  female  26.0      0      0   7.9250   NaN   
3              4         1       1  female  35.0      1      0  53.1000  C123   
4              5         0       3    male  35.0      0      0   8.0500   NaN   
..           ...       ...     ...     ...   ...    ...    ...      ...   ...   
886          887         0       2    male  27.0      0      0  13.0000   NaN   
887          888         1       1  female  19.0      0      0  30.0000   B42   
888          889         0       3  female  28.0      1      2  23.4500   NaN   
889          890         1       1    male  26.0      0      0  30.0000  C148   
890          891         0       3    male  32.0      0      0   7.7500   NaN   

    Embarked Cabin_First  
0          S           n  
1          C           C  
2          S           n  
3          S           C  
4          S           n  
..       ...         ...  
886        S           n  
887        S           B  
888        S           n  
889        C           C  
890        Q           n  

[891 rows x 11 columns]

In [123]:
test

PassengerId  Pclass     Sex   Age  SibSp  Parch      Fare Cabin Embarked  \
0            892       3    male  34.5      0      0    7.8292   NaN        Q   
1            893       3  female  47.0      1      0    7.0000   NaN        S   
2            894       2    male  62.0      0      0    9.6875   NaN        Q   
3            895       3    male  27.0      0      0    8.6625   NaN        S   
4            896       3  female  22.0      1      1   12.2875   NaN        S   
..           ...     ...     ...   ...    ...    ...       ...   ...      ...   
413         1305       3    male  28.0      0      0    8.0500   NaN        S   
414         1306       1  female  39.0      0      0  108.9000  C105        C   
415         1307       3    male  38.5      0      0    7.2500   NaN        S   
416         1308       3    male  28.0      0      0    8.0500   NaN        S   
417         1309       3    male  28.0      1      1   22.3583   NaN        C   

    Cabin_First  
0             n  
1             n  
2             n  
3             n  
4             n  
..          ...  
413           n  
414           C  
415           n  
416           n  
417           n  

[418 rows x 10 columns]

Cabin_yes_train
fare이랑 cabin_frist가 음의 상관관계이므로 비용이 증가하면 좋은 선실이다
cabin_First와 pclass는 상관관계가있따 cabin_first가 낮을수록 pclass는 낮다.

In [124]:
train

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin  \
0              1         0       3    male  22.0      1      0   7.2500   NaN   
1              2         1       1  female  38.0      1      0  71.2833   C85   
2              3         1       3  female  26.0      0      0   7.9250   NaN   
3              4         1       1  female  35.0      1      0  53.1000  C123   
4              5         0       3    male  35.0      0      0   8.0500   NaN   
..           ...       ...     ...     ...   ...    ...    ...      ...   ...   
886          887         0       2    male  27.0      0      0  13.0000   NaN   
887          888         1       1  female  19.0      0      0  30.0000   B42   
888          889         0       3  female  28.0      1      2  23.4500   NaN   
889          890         1       1    male  26.0      0      0  30.0000  C148   
890          891         0       3    male  32.0      0      0   7.7500   NaN   

    Embarked Cabin_First  
0          S           n  
1          C           C  
2          S           n  
3          S           C  
4          S           n  
..       ...         ...  
886        S           n  
887        S           B  
888        S           n  
889        C           C  
890        Q           n  

[891 rows x 11 columns]

In [125]:
test

PassengerId  Pclass     Sex   Age  SibSp  Parch      Fare Cabin Embarked  \
0            892       3    male  34.5      0      0    7.8292   NaN        Q   
1            893       3  female  47.0      1      0    7.0000   NaN        S   
2            894       2    male  62.0      0      0    9.6875   NaN        Q   
3            895       3    male  27.0      0      0    8.6625   NaN        S   
4            896       3  female  22.0      1      1   12.2875   NaN        S   
..           ...     ...     ...   ...    ...    ...       ...   ...      ...   
413         1305       3    male  28.0      0      0    8.0500   NaN        S   
414         1306       1  female  39.0      0      0  108.9000  C105        C   
415         1307       3    male  38.5      0      0    7.2500   NaN        S   
416         1308       3    male  28.0      0      0    8.0500   NaN        S   
417         1309       3    male  28.0      1      1   22.3583   NaN        C   

    Cabin_First  
0             n  
1             n  
2             n  
3             n  
4             n  
..          ...  
413           n  
414           C  
415           n  
416           n  
417           n  

[418 rows x 10 columns]

In [126]:
from sklearn.preprocessing import LabelEncoder
features = ['Cabin_First','Sex', 'Embarked']
for feature in features:
    le = LabelEncoder()
    le = le.fit(train[feature])
    train[feature] = le.transform(train[feature])
    test[feature] = le.transform(test[feature])
# 남자가 0 여자가 1

In [127]:
# from sklearn.preprocessing import LabelEncoder
# features = ['Cabin_First','Sex', 'Embarked']
# for feature in features:
#     le = LabelEncoder()
#     le = le.fit(test[feature])
#     test[feature] = le.transform(test[feature])
# # 남자가 0 여자가 1

In [128]:
train = train.drop(columns=["Cabin"])

In [129]:
train

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  \
0              1         0       3    1  22.0      1      0   7.2500   
1              2         1       1    0  38.0      1      0  71.2833   
2              3         1       3    0  26.0      0      0   7.9250   
3              4         1       1    0  35.0      1      0  53.1000   
4              5         0       3    1  35.0      0      0   8.0500   
..           ...       ...     ...  ...   ...    ...    ...      ...   
886          887         0       2    1  27.0      0      0  13.0000   
887          888         1       1    0  19.0      0      0  30.0000   
888          889         0       3    0  28.0      1      2  23.4500   
889          890         1       1    1  26.0      0      0  30.0000   
890          891         0       3    1  32.0      0      0   7.7500   

     Embarked  Cabin_First  
0           2            8  
1           0            2  
2           2            8  
3           2            2  
4           2            8  
..        ...          ...  
886         2            8  
887         2            1  
888         2            8  
889         0            2  
890         1            8  

[891 rows x 10 columns]

In [130]:
test = test.drop(columns=["Cabin"])

In [131]:
train

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  \
0              1         0       3    1  22.0      1      0   7.2500   
1              2         1       1    0  38.0      1      0  71.2833   
2              3         1       3    0  26.0      0      0   7.9250   
3              4         1       1    0  35.0      1      0  53.1000   
4              5         0       3    1  35.0      0      0   8.0500   
..           ...       ...     ...  ...   ...    ...    ...      ...   
886          887         0       2    1  27.0      0      0  13.0000   
887          888         1       1    0  19.0      0      0  30.0000   
888          889         0       3    0  28.0      1      2  23.4500   
889          890         1       1    1  26.0      0      0  30.0000   
890          891         0       3    1  32.0      0      0   7.7500   

     Embarked  Cabin_First  
0           2            8  
1           0            2  
2           2            8  
3           2            2  
4           2            8  
..        ...          ...  
886         2            8  
887         2            1  
888         2            8  
889         0            2  
890         1            8  

[891 rows x 10 columns]

In [132]:
train["Cabin_First"].value_counts()

Cabin_First
8    687
2     59
1     47
3     33
4     32
0     15
5     13
6      4
7      1
Name: count, dtype: int64

In [133]:
test

PassengerId  Pclass  Sex   Age  SibSp  Parch      Fare  Embarked  \
0            892       3    1  34.5      0      0    7.8292         1   
1            893       3    0  47.0      1      0    7.0000         2   
2            894       2    1  62.0      0      0    9.6875         1   
3            895       3    1  27.0      0      0    8.6625         2   
4            896       3    0  22.0      1      1   12.2875         2   
..           ...     ...  ...   ...    ...    ...       ...       ...   
413         1305       3    1  28.0      0      0    8.0500         2   
414         1306       1    0  39.0      0      0  108.9000         0   
415         1307       3    1  38.5      0      0    7.2500         2   
416         1308       3    1  28.0      0      0    8.0500         2   
417         1309       3    1  28.0      1      1   22.3583         0   

     Cabin_First  
0              8  
1              8  
2              8  
3              8  
4              8  
..           ...  
413            8  
414            2  
415            8  
416            8  
417            8  

[418 rows x 9 columns]

In [134]:
test["Cabin_First"].value_counts()

Cabin_First
8    327
2     35
1     18
3     13
4      9
5      8
0      7
6      1
Name: count, dtype: int64

In [135]:
train["new_age"] = train["Age"].apply(lambda x : int(x/10))
test["new_age"] = test["Age"].apply(lambda x : int(x/10))
train = train.drop(columns=["PassengerId","Age"])
test = test.drop(columns=["Age"])

In [136]:
train

Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  \
0           0       3    1      1      0   7.2500         2            8   
1           1       1    0      1      0  71.2833         0            2   
2           1       3    0      0      0   7.9250         2            8   
3           1       1    0      1      0  53.1000         2            2   
4           0       3    1      0      0   8.0500         2            8   
..        ...     ...  ...    ...    ...      ...       ...          ...   
886         0       2    1      0      0  13.0000         2            8   
887         1       1    0      0      0  30.0000         2            1   
888         0       3    0      1      2  23.4500         2            8   
889         1       1    1      0      0  30.0000         0            2   
890         0       3    1      0      0   7.7500         1            8   

     new_age  
0          2  
1          3  
2          2  
3          3  
4          3  
..       ...  
886        2  
887        1  
888        2  
889        2  
890        3  

[891 rows x 9 columns]

In [137]:
test

PassengerId  Pclass  Sex  SibSp  Parch      Fare  Embarked  Cabin_First  \
0            892       3    1      0      0    7.8292         1            8   
1            893       3    0      1      0    7.0000         2            8   
2            894       2    1      0      0    9.6875         1            8   
3            895       3    1      0      0    8.6625         2            8   
4            896       3    0      1      1   12.2875         2            8   
..           ...     ...  ...    ...    ...       ...       ...          ...   
413         1305       3    1      0      0    8.0500         2            8   
414         1306       1    0      0      0  108.9000         0            2   
415         1307       3    1      0      0    7.2500         2            8   
416         1308       3    1      0      0    8.0500         2            8   
417         1309       3    1      1      1   22.3583         0            8   

     new_age  
0          3  
1          4  
2          6  
3          2  
4          2  
..       ...  
413        2  
414        3  
415        3  
416        2  
417        2  

[418 rows x 9 columns]

# 다중회귀를 이용해서 CARBIN_fIRST 가 8인값대체
# train의 Cabin을 라벨인코딩하기전 nan값 처리하기

In [138]:
train_data = train[train["Cabin_First"] != 8]  # Cabin_First가 결측값이 아닌 데이터
test_data = test[test["Cabin_First"] != 8] 

In [139]:
train.head(1)

Survived  Pclass  Sex  SibSp  Parch  Fare  Embarked  Cabin_First  new_age
0         0       3    1      1      0  7.25         2            8        2

In [140]:
test.head(1)

PassengerId  Pclass  Sex  SibSp  Parch    Fare  Embarked  Cabin_First  \
0          892       3    1      0      0  7.8292         1            8   

   new_age  
0        3

In [141]:
x_train = train_data[['Pclass', 'Sex', 'new_age', 'SibSp', 'Parch', 'Fare', 'Embarked']]  # 예측 변수들
x_test = test_data[['Pclass', 'Sex', 'new_age', 'SibSp', 'Parch', 'Fare', 'Embarked']]

In [142]:
y_train = train_data['Cabin_First']
y_test = test_data['Cabin_First']

In [143]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import LabelEncoder
model = LinearRegression()
model.fit(x_train, y_train)
model1 = LinearRegression()
model1.fit(x_test, y_test)

LinearRegression()

In [144]:
x_pred = train[train['Cabin_First'] ==8][['Pclass', 'Sex', 'new_age', 'SibSp', 'Parch', 'Fare', 'Embarked']]  # 예측에 사용할 데이터

In [145]:
x_pred1 = test[test['Cabin_First'] ==8][['Pclass', 'Sex', 'new_age', 'SibSp', 'Parch', 'Fare', 'Embarked']]  # 예측에 사용할 데이터

In [146]:
predictions = model.predict(x_pred)

In [147]:
predictions1 = model1.predict(x_pred1)

In [148]:
predictions

array([5.51589903, 5.4298045 , 5.26109573, 5.0514925 , 5.9976615 ,
       5.45314693, 3.71087712, 5.2412688 , 5.5483241 , 5.41017999,
       3.93528971, 6.05837968, 3.69620796, 5.69316882, 5.05458804,
       3.67886643, 5.22107073, 6.18583981, 5.73618076, 4.86640973,
       5.24132652, 5.24170967, 1.78564732, 5.24169592, 3.78266344,
       1.86265779, 2.36579116, 4.86639772, 5.2412688 , 5.9258579 ,
       5.29561986, 5.73736967, 4.13385631, 4.86449184, 3.69066171,
       5.2214996 , 5.2412688 , 5.30370244, 5.24169592, 5.36976944,
       5.65408679, 6.21678923, 5.24198358, 4.11956076, 3.89153404,
       4.86639772, 4.10741878, 6.50459542, 4.86639772, 5.99440572,
       1.74599347, 5.13207797, 5.22113223, 6.53186466, 5.78420348,
       3.72318266, 6.69277372, 3.52323171, 5.11808358, 5.12258378,
       5.24170967, 5.2412688 , 3.64332364, 5.43662247, 5.23855264,
       5.23712309, 5.24158871, 2.06780401, 3.87170712, 6.24400179,
       5.46729168, 5.2412688 , 5.2412688 , 5.2412688 , 5.24182

In [149]:
predictions1

array([5.52357157, 5.64317736, 3.64207096, 5.62953602, 5.80453412,
       5.68625378, 5.65077782, 3.94265926, 5.71722238, 5.60020815,
       5.63091497, 2.07295469, 3.66172943, 3.91693131, 3.81047065,
       5.53265395, 5.75697259, 5.54404161, 1.85494974, 5.81027972,
       2.3050384 , 2.08072249, 5.44511225, 5.53265395, 5.50518453,
       3.71945887, 3.88201877, 5.78730689, 5.8398655 , 5.59037584,
       5.75748417, 5.75638256, 5.62802973, 5.54350537, 5.51919613,
       5.51553846, 3.98588187, 5.63086245, 5.58144346, 5.79645152,
       3.81364029, 4.0795607 , 3.81265846, 5.65319097, 5.57318553,
       5.61542289, 5.68864441, 3.85813605, 5.68890664, 5.70833496,
       4.0292229 , 5.76578706, 2.04345844, 5.70828999, 5.63091497,
       5.75770899, 5.63063763, 3.85903533, 5.70828999, 5.68835833,
       2.07295469, 5.63091497, 3.87115277, 5.51891546, 5.70805762,
       5.81521361, 5.70828999, 4.06890945, 5.74622624, 5.63113224,
       5.63063763, 2.08894609, 5.63109483, 5.63086245, 5.75783

In [150]:
len(predictions)

687

In [151]:
len(predictions1)

327

In [152]:
train.isnull().sum()

Survived       0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked       0
Cabin_First    0
new_age        0
dtype: int64

In [153]:
test.isnull().sum()

PassengerId    0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked       0
Cabin_First    0
new_age        0
dtype: int64

# 변수 선택기법

In [154]:
train.loc[train['Cabin_First'] == 8, 'Cabin_First'] = predictions

/tmp/ipykernel_10578/1486097688.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[5.51589903 5.4298045  5.26109573 5.0514925  5.9976615  5.45314693
 3.71087712 5.2412688  5.5483241  5.41017999 3.93528971 6.05837968
 3.69620796 5.69316882 5.05458804 3.67886643 5.22107073 6.18583981
 5.73618076 4.86640973 5.24132652 5.24170967 1.78564732 5.24169592
 3.78266344 1.86265779 2.36579116 4.86639772 5.2412688  5.9258579
 5.29561986 5.73736967 4.13385631 4.86449184 3.69066171 5.2214996
 5.2412688  5.30370244 5.24169592 5.36976944 5.65408679 6.21678923
 5.24198358 4.11956076 3.89153404 4.86639772 4.10741878 6.50459542
 4.86639772 5.99440572 1.74599347 5.13207797 5.22113223 6.53186466
 5.78420348 3.72318266 6.69277372 3.52323171 5.11808358 5.12258378
 5.24170967 5.2412688  3.64332364 5.43662247 5.23855264 5.23712309
 5.24158871 2.06780401 3.87170712 6.24400179 5.46729168 5.2412688
 5.2412688  5.2412688  5.2418286

In [155]:
test.loc[test['Cabin_First'] == 8, 'Cabin_First'] = predictions1

/tmp/ipykernel_10578/4177779283.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[5.52357157 5.64317736 3.64207096 5.62953602 5.80453412 5.68625378
 5.65077782 3.94265926 5.71722238 5.60020815 5.63091497 2.07295469
 3.66172943 3.91693131 3.81047065 5.53265395 5.75697259 5.54404161
 1.85494974 5.81027972 2.3050384  2.08072249 5.44511225 5.53265395
 5.50518453 3.71945887 3.88201877 5.78730689 5.8398655  5.59037584
 5.75748417 5.75638256 5.62802973 5.54350537 5.51919613 5.51553846
 3.98588187 5.63086245 5.58144346 5.79645152 3.81364029 4.0795607
 3.81265846 5.65319097 5.57318553 5.61542289 5.68864441 3.85813605
 5.68890664 5.70833496 4.0292229  5.76578706 2.04345844 5.70828999
 5.63091497 5.75770899 5.63063763 3.85903533 5.70828999 5.68835833
 2.07295469 5.63091497 3.87115277 5.51891546 5.70805762 5.81521361
 5.70828999 4.06890945 5.74622624 5.63113224 5.63063763 2.08894609
 5.63109483 5.63086245 5.75783

# 해보고싶은거!!!! 선택적 변수선택기법

In [156]:
# import pandas as pd
# import statsmodels.api as sm

# # 데이터셋 로드 (예시 데이터셋)
# data = {
#     'PassengerId': range(1, 892),
#     'Survived': [0, 1, 1, 1, 0, 0, 0, 1, 0, 1],
#     'Pclass': [3, 1, 3, 1, 3, 2, 1, 3, 3, 3],
#     'Sex': ['male', 'female', 'female', 'female', 'male', 'male', 'female', 'female', 'male', 'male'],
#     'Age': [22.0, 38.0, 26.0, 35.0, 35.0, 27.0, 19.0, 28.0, 26.0, 32.0],
#     'SibSp': [1, 1, 0, 1, 0, 0, 0, 1, 0, 0],
#     'Parch': [0, 0, 0, 0, 0, 0, 0, 2, 0, 0],
#     'Fare': [7.25, 71.2833, 7.9250, 53.1000, 8.0500, 13.0000, 30.0000, 23.4500, 30.0000, 7.7500],
#     'Embarked': ['S', 'C', 'S', 'S', 'S', 'S', 'S', 'S', 'C', 'Q'],
#     'Cabin_First': ['n', 'C', 'n', 'C', 'n', 'n', 'B', 'n', 'C', 'n']
# }
# df = pd.DataFrame(data)

# # Sex 변수를 이진 변수로 변환 (male: 1, female: 0)
# df['Sex'] = df['Sex'].map({'male': 1, 'female': 0})

# # Embarked 변수를 이진 변수로 변환 (S: 0, C: 1, Q: 2)
# df['Embarked'] = df['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

# # 타겟 변수 및 예측 변수 설정
# y = df['Cabin_First']
# X = df[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']]

# # 단계적 선택법을 사용하여 변수 선택
# def stepwise_selection(X, y, initial_list=[], threshold_in=0.01, threshold_out=0.05, verbose=True):
#     included = list(initial_list)
#     while True:
#         changed = False
#         # 전진 선택법
#         excluded = list(set(X.columns) - set(included))
#         new_pval = pd.Series(index=excluded)
#         for new_column in excluded:
#             model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included + [new_column]]))).fit()
#             new_pval[new_column] = model.pvalues[new_column]
#         best_pval = new_pval.min()
#         if best_pval < threshold_in:
#             best_feature = new_pval.idxmin()
#             included.append(best_feature)
#             changed = True
#             if verbose:
#                 print('Add  {:30} with p-value {:.6}'.format(best_feature, best_pval))

#         # 후진 제거법
#         model = sm.OLS(y, sm.add_constant(pd.DataFrame(X[included]))).fit()
#         pvalues = model.pvalues.iloc[1:]
#         worst_pval = pvalues.max()
#         if worst_pval > threshold_out:
#             changed = True
#             worst_feature = pvalues.idxmax()
#             included.remove(worst_feature)
#             if verbose:
#                 print('Drop {:30} with p-value {:.6}'.format(worst_feature, worst_pval))
#         if not changed:
#             break
#     return included

# selected_features = stepwise_selection(X, y)
# print('Selected features:', selected_features)

# # 선택된 변수로 모델링
# final_model = sm.OLS(y, sm.add_constant(X[selected_features])).fit()
# print(final_model.summary())


In [157]:
numeric_columns = train.select_dtypes(include='number')
correlation_matrix = numeric_columns.corr(method='pearson')
correlation_matrix

# fare이랑 cabin_frist가 음의 상관관계이므로 비용이 증가하면 좋은 선실이다
# cabin_First와 pclass는 상관관계가있따 cabin_first가 낮을수록 pclass는 낮다.

Survived    Pclass       Sex     SibSp     Parch      Fare  \
Survived     1.000000 -0.338481 -0.543351 -0.035322  0.081629  0.257307   
Pclass      -0.338481  1.000000  0.131900  0.083081  0.018443 -0.549500   
Sex         -0.543351  0.131900  1.000000 -0.114631 -0.245489 -0.182333   
SibSp       -0.035322  0.083081 -0.114631  1.000000  0.414838  0.159651   
Parch        0.081629  0.018443 -0.245489  0.414838  1.000000  0.216225   
Fare         0.257307 -0.549500 -0.182333  0.159651  0.216225  1.000000   
Embarked    -0.163517  0.157112  0.104057  0.066654  0.038322 -0.221226   
Cabin_First -0.292985  0.904732  0.055325  0.263509  0.100634 -0.527976   
new_age     -0.050972 -0.339610  0.074794 -0.226973 -0.158376  0.099775   

             Embarked  Cabin_First   new_age  
Survived    -0.163517    -0.292985 -0.050972  
Pclass       0.157112     0.904732 -0.339610  
Sex          0.104057     0.055325  0.074794  
SibSp        0.066654     0.263509 -0.226973  
Parch        0.038322     0.100634 -0.158376  
Fare        -0.221226    -0.527976  0.099775  
Embarked     1.000000     0.256845 -0.003150  
Cabin_First  0.256845     1.000000 -0.326055  
new_age     -0.003150    -0.326055  1.000000

In [158]:
train.isnull().sum()

Survived       0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked       0
Cabin_First    0
new_age        0
dtype: int64

In [159]:
test.isnull().sum()

PassengerId    0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked       0
Cabin_First    0
new_age        0
dtype: int64

### 상관계수
Pclass와 survived -0.338
survived와 sex -0.543
pclass와 age -0.369
sex와 parch -0.245
parch와 sibsp 0.414
fare과 survived 0.25
fare과 pclass -0.90
fare과 cabin_First
survived cabin_First -0.3

In [160]:
correlation_matrix

Survived    Pclass       Sex     SibSp     Parch      Fare  \
Survived     1.000000 -0.338481 -0.543351 -0.035322  0.081629  0.257307   
Pclass      -0.338481  1.000000  0.131900  0.083081  0.018443 -0.549500   
Sex         -0.543351  0.131900  1.000000 -0.114631 -0.245489 -0.182333   
SibSp       -0.035322  0.083081 -0.114631  1.000000  0.414838  0.159651   
Parch        0.081629  0.018443 -0.245489  0.414838  1.000000  0.216225   
Fare         0.257307 -0.549500 -0.182333  0.159651  0.216225  1.000000   
Embarked    -0.163517  0.157112  0.104057  0.066654  0.038322 -0.221226   
Cabin_First -0.292985  0.904732  0.055325  0.263509  0.100634 -0.527976   
new_age     -0.050972 -0.339610  0.074794 -0.226973 -0.158376  0.099775   

             Embarked  Cabin_First   new_age  
Survived    -0.163517    -0.292985 -0.050972  
Pclass       0.157112     0.904732 -0.339610  
Sex          0.104057     0.055325  0.074794  
SibSp        0.066654     0.263509 -0.226973  
Parch        0.038322     0.100634 -0.158376  
Fare        -0.221226    -0.527976  0.099775  
Embarked     1.000000     0.256845 -0.003150  
Cabin_First  0.256845     1.000000 -0.326055  
new_age     -0.003150    -0.326055  1.000000

In [161]:
train

Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  \
0           0       3    1      1      0   7.2500         2     5.515899   
1           1       1    0      1      0  71.2833         0     2.000000   
2           1       3    0      0      0   7.9250         2     5.429804   
3           1       1    0      1      0  53.1000         2     2.000000   
4           0       3    1      0      0   8.0500         2     5.261096   
..        ...     ...  ...    ...    ...      ...       ...          ...   
886         0       2    1      0      0  13.0000         2     3.696208   
887         1       1    0      0      0  30.0000         2     1.000000   
888         0       3    0      1      2  23.4500         2     5.690275   
889         1       1    1      0      0  30.0000         0     2.000000   
890         0       3    1      0      0   7.7500         1     5.073345   

     new_age  
0          2  
1          3  
2          2  
3          3  
4          3  
..       ...  
886        2  
887        1  
888        2  
889        2  
890        3  

[891 rows x 9 columns]

In [162]:
train.isnull().sum()

Survived       0
Pclass         0
Sex            0
SibSp          0
Parch          0
Fare           0
Embarked       0
Cabin_First    0
new_age        0
dtype: int64

In [163]:
train

Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  \
0           0       3    1      1      0   7.2500         2     5.515899   
1           1       1    0      1      0  71.2833         0     2.000000   
2           1       3    0      0      0   7.9250         2     5.429804   
3           1       1    0      1      0  53.1000         2     2.000000   
4           0       3    1      0      0   8.0500         2     5.261096   
..        ...     ...  ...    ...    ...      ...       ...          ...   
886         0       2    1      0      0  13.0000         2     3.696208   
887         1       1    0      0      0  30.0000         2     1.000000   
888         0       3    0      1      2  23.4500         2     5.690275   
889         1       1    1      0      0  30.0000         0     2.000000   
890         0       3    1      0      0   7.7500         1     5.073345   

     new_age  
0          2  
1          3  
2          2  
3          3  
4          3  
..       ...  
886        2  
887        1  
888        2  
889        2  
890        3  

[891 rows x 9 columns]

In [164]:
train

Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  \
0           0       3    1      1      0   7.2500         2     5.515899   
1           1       1    0      1      0  71.2833         0     2.000000   
2           1       3    0      0      0   7.9250         2     5.429804   
3           1       1    0      1      0  53.1000         2     2.000000   
4           0       3    1      0      0   8.0500         2     5.261096   
..        ...     ...  ...    ...    ...      ...       ...          ...   
886         0       2    1      0      0  13.0000         2     3.696208   
887         1       1    0      0      0  30.0000         2     1.000000   
888         0       3    0      1      2  23.4500         2     5.690275   
889         1       1    1      0      0  30.0000         0     2.000000   
890         0       3    1      0      0   7.7500         1     5.073345   

     new_age  
0          2  
1          3  
2          2  
3          3  
4          3  
..       ...  
886        2  
887        1  
888        2  
889        2  
890        3  

[891 rows x 9 columns]

In [165]:
test

PassengerId  Pclass  Sex  SibSp  Parch      Fare  Embarked  Cabin_First  \
0            892       3    1      0      0    7.8292         1     5.523572   
1            893       3    0      1      0    7.0000         2     5.643177   
2            894       2    1      0      0    9.6875         1     3.642071   
3            895       3    1      0      0    8.6625         2     5.629536   
4            896       3    0      1      1   12.2875         2     5.804534   
..           ...     ...  ...    ...    ...       ...       ...          ...   
413         1305       3    1      0      0    8.0500         2     5.630638   
414         1306       1    0      0      0  108.9000         0     2.000000   
415         1307       3    1      0      0    7.2500         2     5.574347   
416         1308       3    1      0      0    8.0500         2     5.630638   
417         1309       3    1      1      1   22.3583         0     5.560107   

     new_age  
0          3  
1          4  
2          6  
3          2  
4          2  
..       ...  
413        2  
414        3  
415        3  
416        2  
417        2  

[418 rows x 9 columns]

# 모델 학습시키기전 test

In [166]:
train

Survived  Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  \
0           0       3    1      1      0   7.2500         2     5.515899   
1           1       1    0      1      0  71.2833         0     2.000000   
2           1       3    0      0      0   7.9250         2     5.429804   
3           1       1    0      1      0  53.1000         2     2.000000   
4           0       3    1      0      0   8.0500         2     5.261096   
..        ...     ...  ...    ...    ...      ...       ...          ...   
886         0       2    1      0      0  13.0000         2     3.696208   
887         1       1    0      0      0  30.0000         2     1.000000   
888         0       3    0      1      2  23.4500         2     5.690275   
889         1       1    1      0      0  30.0000         0     2.000000   
890         0       3    1      0      0   7.7500         1     5.073345   

     new_age  
0          2  
1          3  
2          2  
3          3  
4          3  
..       ...  
886        2  
887        1  
888        2  
889        2  
890        3  

[891 rows x 9 columns]

In [167]:
from sklearn.model_selection import train_test_split
x = train.drop(['Survived'], axis=1)  # Survived를 제외한 모든 열을 설명 변수로 사용

# 종속 변수 (Target)
y = train['Survived']

# train과 test 셋으로 데이터 나누기 (일반적으로 70%는 train, 30%는 test로 나누는 경우가 많습니다)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

In [168]:
x

Pclass  Sex  SibSp  Parch     Fare  Embarked  Cabin_First  new_age
0         3    1      1      0   7.2500         2     5.515899        2
1         1    0      1      0  71.2833         0     2.000000        3
2         3    0      0      0   7.9250         2     5.429804        2
3         1    0      1      0  53.1000         2     2.000000        3
4         3    1      0      0   8.0500         2     5.261096        3
..      ...  ...    ...    ...      ...       ...          ...      ...
886       2    1      0      0  13.0000         2     3.696208        2
887       1    0      0      0  30.0000         2     1.000000        1
888       3    0      1      2  23.4500         2     5.690275        2
889       1    1      0      0  30.0000         0     2.000000        2
890       3    1      0      0   7.7500         1     5.073345        3

[891 rows x 8 columns]

In [169]:
test_input = test[["Pclass",	"Sex",	"SibSp",	"Parch",	"Fare",	"Embarked"	,"Cabin_First",	"new_age"]]

# 의사결정트리

In [170]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score, precision_score, recall_score
dt = DecisionTreeClassifier()

# 탐색할 매개변수들 설정
param_grid = {
    'max_depth': [None, 5, 10, 15,20],  # 트리의 최대 깊이
    'min_samples_split': [2, 5, 10,15],  # 내부 노드를 분할하기 위한 최소 샘플 수
    'min_samples_leaf': [1, 2, 4,8],  # 리프 노드에 필요한 최소 샘플 수
    'max_features': [None, 'sqrt', 'log2']  # 각 분할에서 고려할 특성의 최대 수
}

# Grid Search를 사용하여 최적의 매개변수 탐색
grid_search = GridSearchCV(estimator=dt, param_grid=param_grid, cv=10, scoring='accuracy', verbose=1)
grid_search.fit(x_train, y_train)

# 최적의 매개변수 출력
print("Best Parameters:", grid_search.best_params_)

# 최적의 모델로 평가
best_model = grid_search.best_estimator_
y_pred = best_model.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy on test set: {accuracy:.4f}")
# 정밀도 평가
precision = precision_score(y_test, y_pred)
print(f"Precision on test set: {precision:.4f}")
# 재현율 평가
recall = recall_score(y_test, y_pred)
print(f"Recall on test set: {recall:.4f}")

Fitting 10 folds for each of 240 candidates, totalling 2400 fits
Best Parameters: {'max_depth': 5, 'max_features': None, 'min_samples_leaf': 2, 'min_samples_split': 5}
Accuracy on test set: 0.7799
Precision on test set: 0.7826
Recall on test set: 0.6486


In [171]:
y_pred = best_model.predict(test_input)

In [172]:
y_pred

array([0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [173]:
answer = pd.concat([test['PassengerId'], pd.Series(y_pred, name='Prediction')], axis=1)
answer.rename(columns={"Prediction" : "Survived"},inplace=True)
answer.to_csv("answer.csv", index = False)


In [174]:
answer

PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         0
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [175]:
answer.to_csv

<bound method NDFrame.to_csv of      PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         0
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]>

# 선형회귀모델

In [176]:
test_input

Pclass  Sex  SibSp  Parch      Fare  Embarked  Cabin_First  new_age
0         3    1      0      0    7.8292         1     5.523572        3
1         3    0      1      0    7.0000         2     5.643177        4
2         2    1      0      0    9.6875         1     3.642071        6
3         3    1      0      0    8.6625         2     5.629536        2
4         3    0      1      1   12.2875         2     5.804534        2
..      ...  ...    ...    ...       ...       ...          ...      ...
413       3    1      0      0    8.0500         2     5.630638        2
414       1    0      0      0  108.9000         0     2.000000        3
415       3    1      0      0    7.2500         2     5.574347        3
416       3    1      0      0    8.0500         2     5.630638        2
417       3    1      1      1   22.3583         0     5.560107        2

[418 rows x 8 columns]

In [177]:
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score
import numpy as np

# KFold 교차 검증 설정
kf = KFold(n_splits=5, shuffle=True, random_state=42)  # 5-fold cross validation 설정
fold_idx = 0
best_accuracy = 0
best_fold = 0
model_list = []
for train_idx, test_idx in kf.split(x):
    fold_idx += 1
    print(f"Fold {fold_idx}:")
    x_train, x_test = x.iloc[train_idx], x.iloc[test_idx]
    y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]
    
    # 모델 학습 (예시로 로지스틱 회귀 모델 사용)
    model = LogisticRegression()
    model.fit(x_train, y_train)
    
    # 모델 평가
    y_pred = model.predict(x_test)
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    
    # 출력
    print(f"Accuracy: {accuracy:.4f}, Precision: {precision:.4f}, Recall: {recall:.4f}")
    model_list.append(model)
    # 최고의 성능을 보이는 fold 찾기
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_fold = fold_idx

print(f"\nBest Fold: {best_fold}, Highest Accuracy: {best_accuracy:.4f}")


Fold 1:
Accuracy: 0.7933, Precision: 0.7681, Recall: 0.7162
Fold 2:
Accuracy: 0.7978, Precision: 0.7925, Recall: 0.6269
Fold 3:
Accuracy: 0.8202, Precision: 0.7714, Recall: 0.7714
Fold 4:
Accuracy: 0.7753, Precision: 0.7097, Recall: 0.6667
Fold 5:
Accuracy: 0.7809, Precision: 0.6912, Recall: 0.7231

Best Fold: 3, Highest Accuracy: 0.8202


/home/encore/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/encore/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regressi

In [178]:
model_list[2]

LogisticRegression()

In [179]:
y_pred = model_list[best_fold-1].predict(test_input)

In [180]:
len(y_pred)

418

In [181]:
answer = pd.concat([test['PassengerId'], pd.Series(y_pred, name='Prediction')], axis=1)
answer.rename(columns={"Prediction" : "Survived"},inplace=True)
answer.to_csv("answer.csv", index = False)
